text/0000-add-limits-trait.md

```diff
@@ -0,0 +1,133 @@
+- Feature Name: Limits trait for the rust types
+- Start Date: 2017-12-20
+- RFC PR: 
+- Rust Issue: 
+
+# Summary
+This is an RFC to add a universal trait for the type limits.
+
+# Motivation
+The motivation is quite simple: [make an ability to accept template types with with limits by requiring a trait](https://stackoverflow.com/questions/47904954/rust-finding-the-maximum-allowable-value-for-generic-data-type-t) 
+so that it simplifies and generalizes the code. Another motivation is that we have all that `max_value()` and `min_value()` implemented as
+usual methods of a type implementation, generalizing this to a trait makes the code simplier and avoids duplicating code. Also, looking at
+C++ template of `std::numeric_limits` tells us we must have this thing too because it is easier to use.
+
```

Comment:
Please elaborate on why this has to be in the standard library and not in a separate crate.

Reply:
What about because the standard library types should implement it ? Also, we have it already implicitly by having `min_value()` and `max_value()` method in each primitive numeric type. One more reason is that such a trait is really not enough to be a separate crate.

Are these reasons enough?

Reply:
Historically no, the same way there isn't a trait for `wrapping_add`, `next_power_of_two`, `rotate_left`, etc, all of which are implemented already on the core types without a trait.

Reply:
@scottmcm Do you think it should be in a separate crate? Could you tell why you think so please?

Reply:
I think that moving this to the `num` crate and removing these static methods from the std's types will cause compilation failures for everyone who uses it.

Reply:
_Removing_, sure, but you don't need to remove the `std` implementations to have them on a trait: http://rust-num.github.io/num/num_traits/bounds/trait.Bounded.html

Reply:
Yes, but why do we need the implementations for the standard types then? And also, we talked about requiring `PartialOrd` and not removing these methods will violate this rule.